src/db/pool.cr

```diff
@@ -282,5 +445,66 @@ module DB
         @mutex.lock
       end
     end
+
+    private def sweep_expired_job
+      timer = @sweep_timer
+      return if timer.nil? || (timer <= Time::Span::ZERO)
+
+      @sweep_job_running = true
+
+      spawn do
+        loop do
+          select
+          when @sweep_job_close_channel.receive
+            @sweep_job_running = false
+            break
+          when timeout(timer)
+          end
+
+          sync do
+            now = Time.utc
+
+            # Although not guaranteed, the first elements of @idle
+            # should be the oldest
+            @idle.each do |resource|
+              expire_info = @resource_lifecycle[resource]
+
+              if lifetime_expired?(expire_info, now) || idle_expired?(expire_info, now)
+                resource.close
+                delete(resource)
+              end
+            end
+
+            ensure_minimum_fresh_resources
+
+            # End job if there is no initial pool size and the entire pool has been expired
+            break if !@initial_pool_size && @total.empty?
+          end
+        end
+      end
+    end
+
+    # Ensure there are at least a minimum of @initial_pool_size non-expired resources
+    #
+    # Should be called after each expiration batch
+    private def ensure_minimum_fresh_resources
+      replenish = (@initial_pool_size - (@total.size + @inflight)).clamp(0, @initial_pool_size)
+
+      return if replenish <= 0
+
+      replenish.times do |index|
```

Comment:
The resource replenishment could actually block a checkout/release operation for a really long time depending on how many resources it needs to be built. In most cases it should just be a single one but it can quickly balloon when many resources are discarded manually from outside of the pool. I'm not entirely sure what's the best way to resolve this.

I feel like spawning a new fiber each time is excessive, so maybe it could be dispatched to the already existing background job?

There's also the fact that any non-blocking operation will remove the guarantee of there always being at least `initial_pool_size` fresh resources available...